### ISSUE TITLE
Provide NSIS installer with bdist_nsis command

### ISSUE THREAD

**Issue by xmikos** [xoviat/subzero#55]
Hello,

before I discovered your project, I have started to write my own `bdist_nsis` command for setuptools to generate NSIS installer from PyInstaller frozen apps. I still prefer NSIS installer to MSI. Would you be interested in adding it to *subzero*? If yes, I can adjust the code and make pull request.

**Comment by xmikos** [comment]
But thinking about it some more, maybe it's not ideal (I wanted to support cx_Freeze too, not only PyInstaller).

**Comment by xoviat** [comment]
The entire point of this project is to take advantage of the development work on PyInstaller. Freezing python applications is *really hard*. cx_Freeze can do it in ~50% of cases maybe. PyInstaller can do it in 90% of cases, but that's just because a lot more work has gone into it (I should know, considering that I am a contributor to that project). cx_Freeze used to be better, but now it's not (I suspect because it stayed on bitbucket, rather than github, for so long).

You can submit your own bdist_nsis command if you want. The point of this project is to provide an end-to-end solution in order to save time. It's not part of PyInstaller because that's already complicated enough.

**Comment by xmikos** [comment]
I have one big issue with PyInstaller (in comparison with cx_Freeze): it doesn't support multiple executables in one package.

I have found that there is pull request [#2416](https://github.com/pyinstaller/pyinstaller/pull/2416) with support for multiple executables made by you, but it isn't merged yet and seems to be stalled. Do you plan to finish it?

**Comment by xmikos** [comment]
(this is main reason why I wnated to support cx_Freeze with my *bdist_nsis* too - I have some projects which need multiple executables in on package)

**Comment by xoviat** [comment]
> I have one big issue with PyInstaller (in comparison with cx_Freeze): it doesn't support multiple executables in one package.

Not exactly. Currently PyInstaller duplicates python bytecode between executables, but not binaries in onedir mode (which is what subzero uses). With subzero, you can have multiple executables in on package, but the package won't be as small as it otherwise could be.

**Comment by xoviat** [comment]
> Do you plan to finish it?

The reason why it's "stalled" is that it needs a review from @codewarrior0. There's nothing that I can do to speed that up.

**Comment by xmikos** [comment]
> Not exactly. Currently PyInstaller duplicates python bytecode between executables, but not binaries in onedir mode (which is what subzero uses). With subzero, you can have multiple executables in one package, but the package won't be as small as it otherwise could be.

I didn't know about this, so if I specify multiple gui_scripts / console_scripts in `setup`, it should just work (only generate bigger dir)? If those executables share something with big dependencies (like PyQt), those libraries will be shared (only Python bytecode is duplicated)?

**Comment by xoviat** [comment]
Yeah, the multiple executable thing is in the [test_hello_world](https://github.com/xoviat/subzero/blob/master/tests/hello_world/setup.py). But don't take my word for it; try it and let me know whether it works (it has worked for me, but who knows?)

**Comment by xmikos** [comment]
Ok, I have tried it and it works! But one thing should be fixed: it seems that you are only supporting `console_scripts`, not `gui_scripts`. Can you please also support `gui_scripts` in entry_points?

Also it would be nice if `gui_scripts` were made automatically `windowed=True` and `console_scripts` automatically `windowed=False` (if not overriden by user using `Executable` with different settings explicitly).

**Comment by xoviat** [comment]
I was actually unaware of `gui_scripts`. I'll implement it later today.

**Comment by xmikos** [comment]
Great!

Btw. `windowed=True` option on `Executable` seems not to work, I have successfully built MSI installer, but after starting app from Start menu, console window is still opened.

**Comment by xmikos** [comment]
Also MSI installer doesn't ask user where to install it?

**Comment by xoviat** [comment]
> 
Btw. windowed=True option on Executable seems not to work, I have successfully built MSI installer, but after starting app from Start menu, console window is still opened.

the option is `console=False`. the README was inaccurate, but I have resolved that now.

> Also MSI installer doesn't ask user where to install it?

I removed that option because I thought it wasn't necessary. I will be switching to go-msi in any case to reduce the amount of code involved in this project. If you need that feature, ASAP, then you can edit https://github.com/xoviat/subzero/blob/master/subzero/resources/Product.wxs.
